### PR TITLE
[BEAM-9674] Don't specify selected fields when fetching BigQuery table size

### PR DESCRIPTION
**Pull request by kmjung**
Don't specify the list of selected fields when making a tables.get call to fetch the table size in the BigQueryStorageTableSource in order to avoid "selected fields too long" errors from the server. The selected field list is not used by the tables.get API to compute the table size.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kmjung** [comment]
R: @chamikaramj

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServices.java [hunk 1]
```diff
@@ -101,10 +101,6 @@ JobStatistics dryRunQuery(String projectId, JobConfigurationQuery queryConfig, S
     @Nullable
     Table getTable(TableReference tableRef) throws InterruptedException, IOException;
 
-    @Nullable
```
Comment (aaltay) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServices.java:104]: Is this deleted a public API?

Comment (kmjung) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServices.java:104]: BigQueryServices is a public interface, although it's meant as a private implementation for BigQuery sources and sinks and is not surfaced at the BigQueryIO level. I can restore this functionality and simply stop calling it from BigQueryStorageStreamSource.

Comment (kmjung) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServices.java]: Done.

**Comment by chamikaramj** [comment]
LGTM. Thanks.

**Comment by chamikaramj** [comment]
Run Java PostCommit

**Comment by chamikaramj** [comment]
Run Dataflow ValidatesRunner

**Comment by kmjung** [comment]
I see the issue that's being exposed by the test failures -- the TableRow converter code assumes that all the fields in the BigQuery table schema are also present in the Avro records, which isn't the case here. We'll need to trim the BigQuery schema on the client side after fetching it.

**Comment by kmjung** [comment]
Run Java PostCommit

**Comment by kmjung** [comment]
Run Dataflow ValidatesRunner

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by kmjung** [comment]
@aaltay should this have triggered a re-test by Jenkins? I don't see the results, if so.

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
> @aaltay should this have triggered a re-test by Jenkins? I don't see the results, if so.

It should have. Sometimes it does not. I retried again.

**Comment by kmjung** [comment]
Thanks!

**Comment by kmjung** [comment]
Run Java PostCommit

**Comment by kmjung** [comment]
Run Dataflow ValidatesRunner

**Comment by aaltay** [comment]
Run Java PostCommit

**Comment by aaltay** [comment]
Run Dataflow ValidatesRunner

**Comment by aaltay** [comment]
Run Java PostCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java [hunk 2]
```diff
@@ -149,11 +151,14 @@
       return ImmutableList.of();
     }
 
+    Schema sessionSchema = new Schema.Parser().parse(readSession.getAvroSchema().getSchema());
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java:154]: Not sure how this trims the schema ? Does readSession.getAvroSchema() somehow has a smaller number of fields than targetTable.getSchema() ?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java [hunk 3]
```diff
@@ -1368,19 +1368,14 @@ public void testStreamSourceSplitAtFractionFailsWhenParentIsPastSplitPoint() thr
...
-
+    Table table = new Table().setTableReference(tableRef).setNumBytes(10L).setSchema(TABLE_SCHEMA);
     fakeDatasetService.createTable(table);
 
     CreateReadSessionRequest expectedCreateReadSessionRequest =
         CreateReadSessionRequest.newBuilder()
             .setParent("projects/project-id")
             .setTableReference(BigQueryHelpers.toTableRefProto(tableRef))
             .setRequestedStreams(10)
-            .setReadOptions(
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java:1382]: Is this a feature regression for the storage API based read path ?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java (hunk 2 above)
Comment (kmjung) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java:154]: With this change, we're no longer specifying the list of selected fields to the tables.get call from which the BigQuery schema is taken; as a result, we get the entire table schema back, so we have to trim it on the client side in the case where the client has specified selected fields. The Avro schema is returned as part of the read session and contains only the selected fields, so we use it as the basis for trimming the BigQuery schema.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java (hunk 3 above)
Comment (kmjung) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java:1382]: I'm not sure that I understand your question here. We used to have e2e tests for BigQueryIO with the read API only for the case where the caller specified all fields as selected fields; with this change, now we have two tests -- one which covers the case where no selected fields are covered (which is effectively the same as the case where all fields are specified), and another which covers the case where only a subset of fields are specified.

**Comment by chamikaramj** [comment]
Thanks for the clarifications. LGTM.

**Pull request merged**